src/Microsoft.TemplateEngine.Cli/TemplateResolution/TemplateResolver.cs

```diff
@@ -485,5 +443,197 @@ private static void AddParameterMatchingToTemplates(IReadOnlyCollection<ITemplat
+        /// <returns></returns>
+        private static Func<ITemplateInfo, MatchInfo?> CliNameFilter(string name)
+        {
+            return (template) =>
+            {
+                var groupAwareTemplate = template as TemplateInfoWithGroupShortNames ?? throw new ArgumentException("CliNameFilter filter supports only TemplateInfoWithGroupShortNames templates");
+
+                if (string.IsNullOrEmpty(name))
+                {
+                    return new MatchInfo { Location = MatchLocation.Name, Kind = MatchKind.Partial };
+                }
+
+                int nameIndex = groupAwareTemplate.Name.IndexOf(name, StringComparison.OrdinalIgnoreCase);
+
+                if (nameIndex == 0 && string.Equals(groupAwareTemplate.Name, name, StringComparison.OrdinalIgnoreCase))
+                {
+                    return new MatchInfo { Location = MatchLocation.Name, Kind = MatchKind.Exact };
+                }
+
+                bool hasShortNamePartialMatch = false;
+
+                foreach (string shortName in groupAwareTemplate.GroupShortNameList)
+                {
+                    int shortNameIndex = shortName.IndexOf(name, StringComparison.OrdinalIgnoreCase);
+
+                    if (shortNameIndex == 0 && string.Equals(shortName, name, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return new MatchInfo { Location = MatchLocation.ShortName, Kind = MatchKind.Exact };
+                    }
+
+                    hasShortNamePartialMatch |= shortNameIndex > -1;
+                }
+
+                if (nameIndex > -1)
+                {
+                    return new MatchInfo { Location = MatchLocation.Name, Kind = MatchKind.Partial };
+                }
+
+                if (hasShortNamePartialMatch)
+                {
+                    return new MatchInfo { Location = MatchLocation.ShortName, Kind = MatchKind.Partial };
+                }
+
+                return new MatchInfo { Location = MatchLocation.Name, Kind = MatchKind.Mismatch };
+            };
+        }
+
+        /// <summary>
+        /// In addition to <see cref="ITemplateInfo"/> the class contains <see cref="TemplateInfoWithGroupShortNames.GroupShortNameList"/> property which contains the short names of other templates in the template group.
+        /// The class is used for template filtering using specific <see cref="CliNameFilter(string)"/> filter which takes into account the short names of template group when matching names.
+        /// </summary>
+        private class TemplateInfoWithGroupShortNames : ITemplateInfo, IShortNameList
+        {
+            private TemplateInfoWithGroupShortNames (ITemplateInfo source)
+            {
+                Author = source.Author;
+                Description = source.Description;
+                Classifications = source.Classifications;
+                DefaultName = source.DefaultName;
```

Comment:
Not actionable: We can probably benefit from source generators to help us with such cases in the long run.